### PR TITLE
ui: expose whether the cost-based optimizer was used in stmt page

### PR DESCRIPTION
**Pull request by knz**
Release note (admin ui change): The "Statements" page now reveals
whether a SQL query used the new cost-based optimizer.

cc @andy-kimball

### CODE REVIEW

**Comment by knz** [comment]
@couchand can you help me with the linter failure? There is a failure in the test "coalesces statements with differing distSQL and failed values" but I don't understand what needs to change. Thanks

File: pkg/ui/src/views/statements/statements.spec.tsx [hunk 1]
```diff
@@ -292,6 +295,7 @@ describe("selectStatement", () => {
     assert.equal(result.stats.count.toNumber(), sumCount);
     assert.deepEqual(result.app, [stmtA.key.app]);
     assert.deepEqual(result.distSQL, [false, true]);
+    assert.deepEqual(result.opt, [false, true]);
```
Comment (couchand) [inline pkg/ui/src/views/statements/statements.spec.tsx]: @knz this is the test case that's failing.  The assertion here says we're expecting to have seen this statement executed with and without `opt` set, but the setup hasn't been updated to create both of them.  It's just making all of them with the default `false`.

Using positional parameters on the `makeFingerprint` helper doesn't really make it obvious, but that's what the four different invocations are doing for the `distSQL` and `failed` keys.

File: pkg/server/serverpb/status.proto [hunk 2]
```diff
@@ -629,6 +629,7 @@ message StatementsResponse {
     bool failed = 4;
     int32 node_id = 5 [ (gogoproto.customname) = "NodeID",
                         (gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/roachpb.NodeID"];
+    bool opt = 6;
```
Comment (couchand) [inline pkg/server/serverpb/status.proto]: Oh this is too bad, I bet the other proto now has `opt = 5`, doesn't it?  Let's try to unfork these maybe in the future?

**Review by couchand — APPROVED** [review]
LGTM pending the test fix

File: pkg/server/serverpb/status.proto (hunk 2 above)
Comment (vilterp) [reply pkg/server/serverpb/status.proto]: can we be a little more descriptive here? `optimizer_used` or something?

**Review by vilterp — CHANGES_REQUESTED** [review]

**Review by knz — COMMENTED** [review]
@vilterp I would prefer if you would follow up with a PR of your own where you perform the clean-ups that you like. I don't really feel comfortable in this area of the code. And then if you do look at this, please also rename "distsql" throughout by "distsql_used" (same story). Thanks

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-:-LImHmqOCXynM2qPVRGe:b7gldas)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/serverpb/status.proto, line 632 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-LIkB38SXjRNtGdM6Zr0-r1-632:-LImHMta32o3BKztFbxu:bm3ovnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f3dbe573ce4f3286961664ac2792fa693d07c/pkg/server/serverpb/status.proto#L632)):*
<details><summary><i>Previously, vilterp (Pete Vilter) wrote…</i></summary><blockquote>

can we be a little more descriptive here? `optimizer_used` or something?
</blockquote></details>

@vilterp I'd prefer that the fields remain named after those in roachpb/app_stats.proto.

@couchand why is there a separate proto here in the first place? Why not simply embedding the one in roachpb/app_stats.proto here?
___
*[pkg/ui/src/views/statements/statements.spec.tsx, line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-LIkB38SXjRNtGdM6Zr5-r1-298:-LImHkOH534Kaocf0NcI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f3dbe573ce4f3286961664ac2792fa693d07c/pkg/ui/src/views/statements/statements.spec.tsx#L298)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

@knz this is the test case that's failing.  The assertion here says we're expecting to have seen this statement executed with and without `opt` set, but the setup hasn't been updated to create both of them.  It's just making all of them with the default `false`.

Using positional parameters on the `makeFingerprint` helper doesn't really make it obvious, but that's what the four different invocations are doing for the `distSQL` and `failed` keys.
</blockquote></details>

Done.

**Review by couchand — DISMISSED** [review]
> Reviewed 4 of 7 files at r1, 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28094)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/serverpb/status.proto, line 632 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-LIkB38SXjRNtGdM6Zr0-r1-632:-LIqUYwv2hMOcXaYYpm9:b-pnf5h6) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f3dbe573ce4f3286961664ac2792fa693d07c/pkg/server/serverpb/status.proto#L632)):*
> Why not simply embedding the one in roachpb/app_stats.proto here?

That sounds like the right solution, the copy/paste was just a kludge to not think too much about it.
___
*[pkg/ui/src/views/statements/statements.spec.tsx, line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-LIkB38SXjRNtGdM6Zr5-r1-298:-LIqUxCuBldAtop-djRf:b-q9k2bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f3dbe573ce4f3286961664ac2792fa693d07c/pkg/ui/src/views/statements/statements.spec.tsx#L298)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

I think you want to go the other way, add additional test inputs that set both values of the opt flag, to show that it's also coalesced.

**Review by knz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28094)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/server/serverpb/status.proto, line 632 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-LIkB38SXjRNtGdM6Zr0-r1-632:-LJJAHxD3n3c2xpQayqQ:b-ns008o) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f3dbe573ce4f3286961664ac2792fa693d07c/pkg/server/serverpb/status.proto#L632)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

> Why not simply embedding the one in roachpb/app_stats.proto here?

That sounds like the right solution, the copy/paste was just a kludge to not think too much about it.
</blockquote></details>

Done. PTAL
___
*[pkg/ui/src/views/statements/statements.spec.tsx, line 298 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/28094#-LIkB38SXjRNtGdM6Zr5-r1-298:-LJJ78Qt90YXod9C13Pv:bo3vl6y) ([raw file](https://github.com/cockroachdb/cockroach/blob/476f3dbe573ce4f3286961664ac2792fa693d07c/pkg/ui/src/views/statements/statements.spec.tsx#L298)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I think you want to go the other way, add additional test inputs that set both values of the opt flag, to show that it's also coalesced.
</blockquote></details>

Thanks, done.

**Review by vilterp — APPROVED** [review]
@couchand is out. This LGTM

**Comment by knz** [comment]
thanks!

bors r+

**Comment by knz** [comment]
gah, somehow the ui unit test fails in TC but not on my machine. need to learn how to parse the test failures.

**Comment by vilterp** [comment]
Are you sure the UI tests failed? In TC it looks like the failures are all under `TestLogic`. I saw some failures there on one of my PRs too; I think they might be broken right now.

**Comment by knz** [comment]
nah definitely in the web ui tests:

https://teamcity.cockroachdb.com/viewLog.html?tab=buildLog&buildTypeId=Cockroach_UnitTests_Check&buildId=821709

```
[Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0) selectApps returns all the apps that appear in the statements FAILED
[16:39:01][Test web UI] 	AssertionError: expected [ '(unset)' ] to deeply equal [ Array(3) ]
[16:39:01][Test web UI] 	    at Context.<anonymous> (src/views/statements/statements.spec.tsx:18348:22)
[16:39:01][Test web UI] HeadlessChrome 0.0.0 (Linux 0.0.0): Executed 259 of 280 (3 FAILED) (0 secs / 3.297 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 260 of 280 (3 FAILED) (0 secs / 3.297 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0) selectTotalFingerprints returns the number of statement fingerprints FAILED
[16:39:01][Test web UI] 	AssertionError: expected 1 to equal 3
[16:39:01][Test web UI] 	    at Context.<anonymous> (src/views/statements/statements.spec.tsx:18360:22)
[16:39:01][Test web UI] HeadlessChrome 0.0.0 (Linux 0.0.0): Executed 261 of 280 (4 FAILED) (0 secs / 3.298 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 262 of 280 (4 FAILED) (0 secs / 3.298 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 263 of 280 (4 FAILED) (0 secs / 3.298 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 264 of 280 (4 FAILED) (0 secs / 3.298 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 265 of 280 (4 FAILED) (0 secs / 3.298 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 266 of 280 (4 FAILED) (0 secs / 3.299 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0): Executed 267 of 280 (4 FAILED) (0 secs / 3.3 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0) selectStatement returns the statement currently loaded FAILED
[16:39:01][Test web UI] 	TypeError: Cannot read property 'count' of undefined
[16:39:01][Test web UI] 	    at Context.<anonymous> (src/views/statements/statements.spec.tsx:18406:41)
[16:39:01][Test web UI] HeadlessChrome 0.0.0 (Linux 0.0.0): Executed 268 of 280 (5 FAILED) (0 secs / 3.3 secs)
[16:39:01][Test web UI]  [1A [2KHeadlessChrome 0.0.0 (Linux 0.0.0) selectStatement coalesces statements from different apps FAILED
[16:39:01][Test web UI] 	TypeError: Cannot read property 'count' of undefined
[16:39:01][Test web UI] 	    at Context.<anonymous> (src/views/statements/statements.spec.tsx:18422:41)
... (more) ...
```

if you could help figuring this out that would be swell

**Comment by knz** [comment]
I had a typo. d'oh.

bors r+

**Comment by couchand** [comment]
@Amruta-Ranade you might add this as well to cockroachdb/docs#3505

**Pull request merged**